src/stats/man/features.1
.TH FEATURES 1 "March 10, 1989" "Copyright 1989 Gary Perlman" "|STAT" "|STAT User's Manual" 
.\" Runoff  $Compile: psroff -man %f
.\" RCS     $Header: /home/cvs/src/stat/man/features.1,v 1.1.1.1 2000/03/18 20:24:22 rich Exp $ 
.SH NAME
features \- Tabulate Features of Items
.SH USAGE
.I features 
[\-h]
[\-c\ col-delim]
[\-f\ inputfile]
[\-i\ itemwidth]
[\-l\ leader]
[\-n\ no-str]
[\-p\ pagesize]
[\-w\ featwidth]
[\-y\ yes-str]
feature[=width]\ [feature[=width]\ ...]
.SH DESCRIPTION
Think of this program as a consumer product comparison checklist tool.
For each item, there are features that you would like the
product to have,
and for some features, several values are possible.
.SH OPTIONS
.de OP\" option-letter [optional value]
.TP
.B -\\$1 \\$2
..
.TP
.B feature
Feature Names.
Feature names indicate which features to tabulate, and in which order,
to describe items.
Features on the command line can be exact matches for item's features,
but to simplify the format of feature files,
features in files can be abbreviated and upper and lower case letters
are treated as equivalent.
For example, Usability on the command line matches usab=good in
the feature file.
You can specify the width of a column for a feature by putting
a width after an equal sign.
See the examples.
.OP c col-delim
Column Delimiter String.
Default: bar (|).
This string is placed between every feature column describing an item.
It can be wider than one character; for example, "\ |\ " would give a
wider table with more space in columns.
.OP f inputfile
Input file.
Default: standard input.
This option allows you to specify a file name for the item description lines.
.OP h
Use Horizontal Format for Headers.
If this option is selected, then all field widths for features must
be specified.
.OP i itemwidth
Width of Items.
Default: 15 characters wide.
If you have ones that are wider, the tables won't line up correctly.
This option allows you to widen the space used to show the items.
.OP l leader
Leader Character After Item Name.
Default: underbar (_).
.OP n no-str
String to Print When Item Feature is Not Present.
Default: space.
.OP p pagesize
Size of Pages.
Default: 60 lines.
For every page, a new feature-table header is generated.
.OP w feat-width
Width of Feature Columns.
Default: 1 character.
This option will allow you to set all the feature column widths.
Individual column widths can be overridden with feature widths
on the command line.
.OP y yes-string
String to Print When Item Has a Feature.
Default: sharp (#).
This can be changed to any character.
It can also be changed for an individual item-feature pair
by using a feature=value format.
For example,
.ce
Perlman eyes=H
could be used to indicate that Perlman's eyes are hazel.
For longer strings,
you will want to adjust the width of the columns
with the -w option or with the feature name on the command line.
.br
.ne 1i
.SH EXAMPLES
.SS "Product Comparisons
.PP
Suppose you want to buy a toaster with a bunch of features like inexpensive,
reliable, wood-grain and one that holds four slices of bread.
These features might be:
price (L=low, M=medium, H=high),
reliability (yes, no),
finish (M=metal, W=wood-grain, O=other),
and
capacity (number of slices of bread).
A set of toasters (items) might be described by a list of features:
.nf
.ta .5i +1i +1i
.ft CW
	BrandX	price=Med reliable finish=Wood capacity=4
	BrandY	price=High finish=Metal capacity=2
	BrandZ	price=Low finish=Wood capacity=4
.ft P
The above input, along with the feature labels:
.ft P
	Price Reliable Finish Capacity
.ft P
in the command:
.ft P
	features -c " | " Price=4 Reliable Finish=6 Capacity
.ft P
will display:
.ft P
	                | Price
	                |      | Reliable
	                |      |   | Finish
	                |      |   |        | Capacity
	BrandX_________ | Med  | # | Wood   | 4 | 
	BrandY_________ | High |   | Metal  | 2 | 
	BrandZ_________ | Low  |   | Wood   | 4 | 
.ft P
.sp
Another command shows how the same information can be viewed differently.
.ft P
	features -l. -c" " -h -nNo -yYes Price=6 Reliable=10 Finish=6 Capacity=10

	                Price  Reliable   Finish Capacity   
	BrandX......... Med    Yes        Wood   4          
	BrandY......... High   No         Metal  2          
	BrandZ......... Low    No         Wood   4          
.ft R
.fi
.SS "Other Uses
The program has been useful for organizing lists of publications
by the topics they cover and for listing the research interests of faculty
in a large Midwest computer science department.
.SH "SEE ALSO"
ptx(1), sort(1)
.SH AUTHOR
Gary Perlman
.SH WARNINGS
Try to avoid similar feature names,
especially ones that begin with the same characters.


src/stats/src/features.c
#include "stat.h"
PGM(features,Tabulate Features of Items,3.1,02/12/90)
/*HISTORY
	03/10/89	3.0	initial version
	02/12/90	3.1	incorporated stat.h
*/

/*
	Special codes such as Consumer Reports advantages and disadvantages
		can be done by:
			Item   Advantages=A,B,C  Disadvantages=a,h
		and then inserting the meanings of the codes at the end of the table.
	Tabs can be inserted between columns with the -c option.
*/

Boole	InfoVersion;          /* print version information */
Boole	InfoLimits;           /* print program limits */
Boole	InfoOptions;          /* print usage information */

#ifdef __STDC__
int printheader (void);
void getfeatures (int argc, char **argv, int optind);
void initial (int argc, char **argv);
int substring (char *s1, char *s2);
int begins (char *s1, char *s2);
int printtable (char *filename);
void printoptions (int argc, char **argv);
void printfield (int fieldno, char *contents);
#else
int printheader ();
void getfeatures ();
void initial ();
int substring ();
int begins ();
int printtable ();
void printoptions ();
void printfield ();
#endif

#define	MAXFEAT    100           /* max number of features */
char	*HasFeature[MAXFEAT];    /* string to indicate having feature */
int 	Featwidth[MAXFEAT];      /* width of each feature column */
char	**Feature;               /* names of features, from cmd line */
int 	Nfeatures;               /* number of features */

char	*Infile = "-";           /* input file name ("-" -> stdin) */

int 	Itemwidth = 15;          /* width of leading field */
int 	Width = 1;               /* width of features */
int 	Pagesize  = 60;          /* lines per page */
char	*Coldelim = "|";         /* used to delimit feature columns */
char	Leader    = '_';         /* leader from end of leading field */
char	*Featyes   = "#";        /* code for having a feature */
char	*Featno    = " ";        /* code for non-membership */
#define	ASSIGN      '='          /* assigned code */
int 	Horizontal = 0;          /* use a one line horizontal header */

main (argc, argv) char **argv;
	{
	extern	int 	optind;
	
	ARGV0;
	initial (argc, argv);
	getfeatures (argc, argv, optind);
	if (printtable (Infile))
		fprintf (stderr, "%s: Can't read from file: '%s'\n", argv[0], Infile);
	exit (0);
	}

/*FUNCTION getfeatures: get features from the command line or file */
void
getfeatures (argc, argv, optind) char **argv;
	{
	char	*ptr;
	Feature = argv + optind;
	Nfeatures = argc - optind;
	
	/* check for width parameters for features */
	for (optind = 0; optind < Nfeatures; optind++)
		{
		if (ptr = strchr (Feature[optind], ASSIGN))
			{
			if (!isinteger (ptr+1))
				ERRNUM(ptr+1,individual column width)
			Featwidth[optind] = atoi (ptr+1);
			*ptr = '\0';
			}
		}
	}

/*FUNCTION initial: initialize program by setting options */
void
initial (argc, argv) char **argv;
	{
	extern	int 	optind;
	extern	char	*optarg;
	char	*optstring = "LOVhc:f:i:l:n:p:w:y:";
	int 	c;
	int 	opterr = 0;
	while ((c = getopt (argc, argv, optstring)) != EOF)
		{
		switch (c)
			{
			case '?':
				opterr++;
				break;
			case 'O': InfoOptions = TRUE; break;
			case 'V': InfoVersion = TRUE; break;
			case 'L': InfoLimits = TRUE; break;
			case 'c': /* column delimiter */
				Coldelim = optarg;
				break;
			case 'f': /* input file */
				Infile = optarg;
				break;
			case 'h': /* horizontal header */
				Horizontal = 1;
				break;
			case 'i': /* item width */
				if (setint (Argv0, c, optarg, &Itemwidth, 1, 100))
					opterr++;
				break;
			case 'l': /* leader character */
				if (*optarg)
					Leader = *optarg;
				break;
			case 'n': /* feature "no" indicator */
				Featno = optarg;
				break;
			case 'p': /* page size */
				if (setint (Argv0, c, optarg, &Pagesize, 10, 200))
					opterr++;
				break;
			case 'w': /* feature column width */
				if (setint (Argv0, c, optarg, &Width, 1, 100))
					opterr++;
				break;
			case 'y': /* yes string */
				Featyes = optarg;
				break;
			}
		}
	if (opterr)
		USAGE ("[-h] [-c col-delim] [-f inputfile] [-i itemwidth]\n\t[-l leader] [-n no-str] [-p pagesize] [-w feat-width] [-y yes-str]")
	usinfo ();
	}

/*FUNCTION getfirst: returns the first field on a line */
char *
getfirst (line)
char	*line;
	{
	static char buffer[100];
	char	*ptr = buffer;
	while (isspace (*line))
		line++;
	while (*line && !isspace (*line))
		*ptr++ = *line++;
	*ptr = '\0';
	if (ptr == buffer)
		return (NULL);
	return (buffer);
	}

/*FUNCTION getrest: returns the rest of a line after the first field */
char *
getrest (line)
char	*line;
	{
	while (isspace (*line))
		line++;
	while (*line && !isspace (*line))
		line++;
	return (line);
	}

/*FUNCTION substring: true if first string is a substring of second */
int
substring (s1, s2)
char	*s1, *s2;
	{
	if (s1 == NULL || *s1 == '\0')
		return (1);
	if (s2 == NULL || *s2 == '\0')
		return (0);
	while (*s2)
		if (begins (s1, s2))
			return (1);
		else
			s2++;
	return (0);
	}

/*FUNCTION begins: true if first string begins the second */
int
begins (s1, s2)
char	*s1, *s2;
	{
	if (s1 == NULL)
		return (1);
	if (s2 == NULL)
		return (0);
	while (*s1)
		{
#		define	makelower(c)  (isupper(c)?tolower(c):(c))
		if (*s1 != *s2 && makelower (*s1) != makelower (*s2))
			return (0);
		s1++;
		s2++;
		}
	return (1);
	}

/*FUNCTION printtable: print the contents of a feature table */
int
printtable (filename)
char	*filename;
	{
	int 	i;
	FILE	*ioptr;
	int 	lineno = Pagesize;     /* will trigger new header */
	char	*item;                 /* the first thing on the line */
	char	*lineptr, *featptr;
	char	line[BUFSIZ];          /* the input line */
	char	*itemvalue;            /* "red" in color=red */

	if (strcmp (filename, "-") == 0)
		{
		ioptr = stdin;
		checkstdin();
		}
	else if ((ioptr = fopen (filename, "r")) == NULL)
		return (1);
	while (fgets (line, sizeof (line), ioptr))
		{
		lineptr = getrest (line);
		if (*lineptr == '\0')
			continue;
		if (lineno == Pagesize)
			lineno = printheader ();
		lineno++;
		item = getfirst (line);
		printf ("%s", item);
		for (i = strlen (item); i < Itemwidth; i++)
			putchar (Leader);
		for (i = 0; i < Nfeatures; i++)
			HasFeature[i] = Featno;
		while (featptr = getfirst (lineptr))
			{
			/*	First strip off item value from any name=value pair
				to will allow the line item to match the Feature name. */
			if (itemvalue = strchr (featptr, ASSIGN))
				*itemvalue++ = '\0';
			else
				itemvalue = Featyes;
			/* Now check for substrings, stop on first match.  */
			for (i = 0; i < Nfeatures; i++)
				{
				if (substring (featptr, Feature[i]))
					{
					if (itemvalue != Featyes)
						itemvalue = strdup (itemvalue);
					HasFeature[i] = itemvalue;
					break;
					}
				}
			lineptr = getrest (lineptr);
			}
		for (i = 0; i < Nfeatures; i++)
			{
			printfield (i, HasFeature[i]);
			if (HasFeature[i] != Featyes && HasFeature[i] != Featno)
				free (HasFeature[i]);
			}
		printf ("%s\n", Coldelim);
		}
	if (ioptr && ioptr != stdin)
		fclose (ioptr);
	return (0);
	}

/*FUNCTION printheader: print a header for a feature table */
/*DESCRIPTION
	This routine prints the Feature labels, at staggered columns
	to create a compact form.
*/
int              /* return number of lines in header */
printheader ()
	{
	int	i, j;
	if (Horizontal)
		{
		for (j = 0; j < Itemwidth; j++)
			putchar (' ');
		for (j = 0; j < Nfeatures; j++)
			printfield (j, Feature[j]);
		printf ("%s\n", Coldelim);
		}
	else
		for (i = 0; i < Nfeatures; i++)
			{
			for (j = 0; j < Itemwidth; j++)
				putchar (' ');
			for (j = 0; j < i; j++)
				printfield (j, " ");
			printf ("%s%s\n", Coldelim, Feature[i]);
			}
	return (Horizontal ? 1 : Nfeatures);
	}

/*FUNCTION printfield: print a field padding or truncating specified contents */
void
printfield (fieldno, contents)
int 	fieldno;     /* the field number to print */
char	*contents;   /* what to place in the field */
	{
	int 	width;
	printf ("%s", Coldelim);
	if (contents == NULL)
		contents = "";
	if (Featwidth[fieldno] || Width != 1)
		{
		if (Featwidth[fieldno])
			width = Featwidth[fieldno];
		else
			width = Width;
		printf ("%-*.*s", width, width, contents);
		}
	else
		printf ("%s", contents);
	}

usinfo ()
	{
	if (InfoVersion)
		pver (Version);
	if (InfoLimits)
		{
		plim (Argv0);
		const (BUFSIZ, "maximum number of characters in lines");
		const (MAXFEAT, "maximum number of features");
		}
	if (InfoOptions)
		{
		ppgm (Argv0, Purpose);
		sopt ('c', "colchar", "Column Separator String", Coldelim);
		sopt ('f', "input", "Input File Name", Infile);
		iopt ('i', "width", "Item Label Width", Itemwidth);
		lopt ('h', "Horizontal Format for Header", Horizontal);
		copt ('l', "leader", "Leader Character After Label", Leader);
		sopt ('n', "no", "String Indicating 'NO'", Featno);
		iopt ('p', "lines", "Page Size", Pagesize);
		iopt ('w', "width", "Width of Feature Columns", Width);
		sopt ('y', "yes", "String Indicating 'YES'", Featyes);
		}
	if (InfoVersion || InfoLimits || InfoOptions)
		exit (0);
	}
